group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -506,32 +555,54 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             .setClientHost(clientHost)
             .build();
 
+        boolean updatedMemberSubscriptions = false;
         if (!updatedMember.equals(member)) {
             records.add(newMemberSubscriptionRecord(groupId, updatedMember));
 
             if (!updatedMember.subscribedTopicNames().equals(member.subscribedTopicNames())) {
                 log.info("[GroupId " + groupId + "] Member " + memberId + " updated its subscribed topics to: " +
                     updatedMember.subscribedTopicNames());
+                updatedMemberSubscriptions = true;
+            }
 
-                subscriptionMetadata = group.computeSubscriptionMetadata(
-                    member,
-                    updatedMember,
-                    topicsImage
-                );
-
-                if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
-                    log.info("[GroupId " + groupId + "] Computed new subscription metadata: "
-                        + subscriptionMetadata + ".");
-                    records.add(newGroupSubscriptionMetadataRecord(groupId, subscriptionMetadata));
-                }
+            if (!updatedMember.subscribedTopicRegex().equals(member.subscribedTopicRegex())) {
+                log.info("[GroupId " + groupId + "] Member " + memberId + " updated its subscribed regex to: " +
+                    updatedMember.subscribedTopicRegex());
+                updatedMemberSubscriptions = true;
+            }
+        }
 
-                groupEpoch += 1;
-                records.add(newGroupEpochRecord(groupId, groupEpoch));
+        long currentTimeMs = time.milliseconds();
+        boolean maybeUpdateMetadata = updatedMemberSubscriptions || group.refreshMetadataNeeded(currentTimeMs);
+        boolean updatedSubscriptionMetadata = false;
+        if (maybeUpdateMetadata) {
+            subscriptionMetadata = group.computeSubscriptionMetadata(
+                member,
+                updatedMember,
+                metadataImage.topics()
+            );
 
-                log.info("[GroupId " + groupId + "] Bumped group epoch to " + groupEpoch + ".");
+            if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
+                log.info("[GroupId " + groupId + "] Computed new subscription metadata: "
+                    + subscriptionMetadata + ".");
+                records.add(newGroupSubscriptionMetadataRecord(groupId, subscriptionMetadata));
+                updatedSubscriptionMetadata = true;
             }
         }
 
+        if (updatedMemberSubscriptions || updatedSubscriptionMetadata) {
+            groupEpoch += 1;
+            records.add(newGroupEpochRecord(groupId, groupEpoch));
+            log.info("[GroupId " + groupId + "] Bumped group epoch to " + groupEpoch + ".");
+        }
+
+        if (maybeUpdateMetadata) {
```

Comment:
While I am able to follow the logic and understand it is probably the best way to avoid duplicate code, I do wonder if there is a less confusing way to express this without the similar booleans. It may not be possible, but maybe we can add some comments.

Reply:
Reworked this part. Let me know if it looks better.